Build Clipper with `tools=no` and patch it to auto-disable exceptions

Xrayez opened the pull request:
Prerequisite for #28987.
Reverts #17319.
~Helps #25262 (needs testing) seems to be fixed for Clipper but other thirdparty code with similar problem detected (assimp), which also uses exceptions.~ already fixed by #29032.

Xrayez commented:
Ok, got this tested and works with sprite editor plugin as before, works alright with #28987 in both `release_debug` and `release` targets under Windows. I can squash commits together but as it's touching thirdparty code, waiting for review. "Patch" is generated with `git diff` against working tree.

akien-mga commented on editor/SCsub:
```diff
@@ -81,7 +81,6 @@ if env['tools']:
     env.add_source_files(env.editor_sources, "*.cpp")
     env_thirdparty = env.Clone()
     env_thirdparty.disable_warnings()
-    env_thirdparty.add_source_files(env.editor_sources, ["#thirdparty/misc/clipper.cpp"])
```
The whole `env_thirdparty` is unused now.

Xrayez replied:
Hmm, well then it's strange why it compiles without linking errors.

akien-mga replied:
Well I mean it does nothing, you remove the only file that was built within `env_thirdparty` which is created two lines above. So you should remove it.

akien-mga commented on thirdparty/misc/clipper.cpp:
```diff
@@ -4329,10 +4336,10 @@ double DistanceFromLineSqrd(
-  //(y� - y�)x + (x� - x�)y + (y� - y�)x� - (x� - x�)y� = 0
-  //A = (y� - y�); B = (x� - x�); C = (y� - y�)x� - (x� - x�)y�
-  //perpendicular distance of point (x�,y�) = (Ax� + By� + C)/Sqrt(A� + B�)
+  //given 2 points (x¹,y¹) & (x²,y²) is ...
```
Encoding change?

Xrayez replied:
This is weird one because `git diff` makes encoding look like this, but the encoding could still be preserved while committing changes... so I left it this way.

Setting `LESSCHARSET=UTF-8 git diff` doesn't seem to cure this either.

akien-mga replied:
`git add -p` should let you selectively add changes and thus ignore those if they're generated by your OS and you can't find how to fix them (I've had similar issues in the past too with weird Mac line endings).

Xrayez replied:
Used `git add -p`, `VSCode` seems to mess up encoding. 🐛

akien-mga replied:
Talking about tools bug, it's weird that your `.patch` file appears as a binary blob in the GitHub diff. In my experience GitHub used to recognize those as text files, but maybe only when created on Unix? See e.g. https://github.com/godotengine/godot/pull/27929/commits/668439d16a5c30587ab77461fb3211306ad62587

Xrayez replied:
I checked the encoding for those, and the patch in 668439d appears as UTF-8 LF, while the patch generated in Windows Git generates UTF-16LE CRLF, but maybe because it's created locally and wasn't checked out from remote yet.

akien-mga replied:
#29098 should fix the encoding issue with these comments, you might have to fight git a bit when rebasing, but hopefully it should be fine since your PR no longer changes this code chunk.

I also pushed a change to enforce LF line endings in the git repo (converted automatically to CRLF when checking out on Windows, and converted back to LF when committing) for all text files, including `.patch` files, so this might solve the diff issue.

Xrayez replied:
Yeah there's plenty of changes so far, I'm going to rebase once it's merged.

akien-mga commented on thirdparty/misc/clipper.hpp:
```diff
@@ -388,17 +388,6 @@ class ClipperOffset
 };
 //------------------------------------------------------------------------------
 
-class clipperException : public std::exception
```
I guess you can also remove `#include <stdexcept>`?

akien-mga commented:
I'd prefer not to have to modify upstream code at all, unless we really need to. In this case, this should ideally be implemented upstream in a way that lets us disable exceptions, similar to https://github.com/nlohmann/json/blob/ee4028b8e4cf6a85002a296a461534e2380f0d57/single_include/nlohmann/json.hpp#L518-L530

But as there's little hope of seeing any new Clipper 6.x release, I guess we don't have much choice. Commits should be squashed, either all together or keeping the revert separate; but given that this revert by itself reintroduces a bug, I guess it's better to include it together with the rest of the changes which prevent the bug from happening.

Xrayez commented:
Angus Johnson is currently working on Clipper 10.0.0. Clipper was written in Delphi originally and translated to C++ pretty much mechanically as the author admits that his C++ knowledge is somewhat limited, so it could take quite some time to propagate changes from upstream for such C++ specific change that he has little knowledge of. I've participated in speeding up Clipper 10.0.0 C++ translation as well so he could delegate this work to me (which would be easier for me just to patch this manually? 😄).

Xrayez commented:
Not related to 6.4.2 but I've tried to translate these exception fixes ideas already to Angus Johnson [here](https://sourceforge.net/p/polyclipping/bugs/192/#e916/cd04/5e97/3bd0/6b4b/65c4/eb0b/4004/f30a/e869/faae).

Commit:
Build Clipper with `tools=no` and patch it to auto-disable exceptions

Reverts "Build polygon clipper only in tools builds" (see #17319)
which allows to build Clipper with tools disabled (release) and because
of that, Clipper has to be patched to optionally disable exceptions in
order to be built on some platforms.

Patched Clipper 6.4.2 to be compiled with exceptions enabled/disabled.
and ensure that Clipper-specific exception macros are defined: don't use
exceptions by default unless exception handling is detected.

Compilation with exceptions will be determined by various
C++ exceptions defines:

* ` __cpp_exceptions` is part of C++ feature testing macros (since C++98);
* `__EXCEPTIONS` is used by some GNU compilers;
* `_CPPUNWIND` is used by MSVC.

The user can override specific exceptions behavior via corresponding
`*_USER` macros (i.e. compiling for embedded systems).

Xrayez commented:
I proposed preliminary changes to upstream here with the patch anyway: 
https://sourceforge.net/p/polyclipping/bugs/193/

akien-mga commented:
Thanks!
